tests/5.0/target/test_target_teams_reduction.c

```diff
@@ -0,0 +1,50 @@
+//===------test_target_teams_reduction.c--------------------------------------===//
+//
+// OpenMP API Version 5.0 Nov 2018
+//
+// Description:
+// This is a basic test to demonstrate target teams + reduction
+//
+//===------------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1024
+
+/**
+  This is a basic test to demonstrate how reduction clause is used
+  with target teams construct.
+*/
+int testTargetTeamsReduction() {
+  int a[N], sum_dev = 0, total = 0;
+  int errors = 0;
+  // Data Inititalize
+  for (int i = 0; i < N; i++) {
+    a[i] = i;
+  }
+  // Execute on target
+#pragma omp target teams map(to: a[0:N]) map(tofrom: sum_dev) reduction(+:total)
+  {
+    for (int i = 0; i < N; i++) {
+      total = total + a[i];
+    }
+    sum_dev = total;
+  }
+  // Validate
+  int sum_host = 0;
+  for (int i = 0; i < N; i++) {
+    sum_host = sum_host + a[i];
+  }
+  OMPVV_TEST_AND_SET_VERBOSE(errors, sum_dev != sum_host);
```

Comment:
This test is correct but does not fully test the teams reduction behavior.

OpenMP Verion 5.0 section 2.19.5.4 reduction Clause from page 300 line 30 to page 301 line 6 says the following:

> For the teams construct, one or more private copies of each list item are created for the initial task of each team in the league, as if the private clause had been used. 
> ...
> At the end of a region that corresponds to a construct for which the reduction clause was specified, the original list item is updated by combining its original value with the final value of each of the private copies, using the combiner of the specified reduction-identifier.
> 

To check the final reduction behavior of the teams reduction clause, how about changing the code as following:

```
  int num_actual_teams = 0; 
  // Execute on target
  #pragma omp target teams map(to: a[0:N]) map(tofrom: num_actual_teams) reduction(+:total)
  for (int i = 0; i < N; i++) {
      total = total + a[i];
      num_actual_teams = omp_get_num_teams();
  }    
  // Validate
  int sum_host = 0; 
  for (int i = 0; i < N; i++) {
      sum_host = sum_host + a[i];
  }    
  OMPVV_TEST_AND_SET_VERBOSE(errors, total != num_actual_teams * sum_host);
```